packages/front-end/components/Settings/EditDataSource/ExperimentAssignmentQueries/AddEditExperimentAssignmentQueryModal.tsx

```diff
@@ -0,0 +1,188 @@
+import React, { FC } from "react";
+import Modal from "../../../Modal";
+import {
+  DataSourceInterfaceWithParams,
+  ExposureQuery,
+} from "back-end/types/datasource";
+import { useForm } from "react-hook-form";
+import cloneDeep from "lodash/cloneDeep";
+import uniqId from "uniqid";
+import Field from "../../../Forms/Field";
+import CodeTextArea from "../../../Forms/CodeTextArea";
+import Tooltip from "../../../Tooltip";
+import Toggle from "../../../Forms/Toggle";
+import StringArrayField from "../../../Forms/StringArrayField";
+
+type EditExperimentAssignmentQueryProps = {
+  exposureQuery?: ExposureQuery;
+  dataSource: DataSourceInterfaceWithParams;
+  mode: "add" | "edit";
+  onSave: (exposureQuery: ExposureQuery) => void;
+  onCancel: () => void;
+};
+
+export const AddEditExperimentAssignmentQueryModal: FC<EditExperimentAssignmentQueryProps> = ({
+  exposureQuery,
+  dataSource,
+  mode,
+  onSave,
+  onCancel,
+}) => {
+  const modalTitle =
+    mode === "add"
+      ? "Add an Experiment Assignment query"
+      : `Edit ${exposureQuery.name}`;
+
+  const userIdTypeOptions = dataSource.settings.userIdTypes.map(
+    ({ userIdType }) => ({
+      display: userIdType,
+      value: userIdType,
+    })
+  );
+  const defaultUserId = userIdTypeOptions[0]?.value || "user_id";
+
+  const form = useForm<ExposureQuery>({
+    defaultValues:
+      mode === "edit"
+        ? cloneDeep<ExposureQuery>(exposureQuery)
+        : {
+            description: "",
+            id: uniqId("tbl_"),
+            name: "",
+            dimensions: [],
+            query: `SELECT\n  ${defaultUserId} as ${defaultUserId},\n  timestamp as timestamp,\n  experiment_id as experiment_id,\n  variation_id as variation_id\nFROM my_table`,
+            userIdType: userIdTypeOptions[0]?.value || "",
+          },
+  });
+
+  // User-entered values
+  const userEnteredUserIdType = form.watch("userIdType");
+  const userEnteredQuery = form.watch("query");
+  const userEnteredHasNameCol = form.watch("hasNameCol");
+  const userEnteredDimensions = form.watch("dimensions");
+
+  const handleSubmit = form.handleSubmit(async (value) => {
+    onSave(value);
+
+    form.reset({
+      id: null,
+      query: "",
+      name: "",
+      dimensions: [],
+      description: "",
+      hasNameCol: false,
+      userIdType: null,
+    });
```

Comment:
Since `onSave` is not awaited, it will wipe all the fields and close the modal immediately.  If there's an error in the API call, the user won't know about it.

I think this is true of all the editing sections, although I'm just catching it now.

Reply:
Good catch. I'll go through all these and fix that, thanks for looking so closely and catching that.